File: frontend/apis/leetcode.ts
import axios from 'axios';
import {
  LeetcodeQuestion,
  LeetcodeQuestionFilters,
  LeetcodeQuestionMetadata,
  leetcodeQuestion,
  leetcodeQuestionMetadata,
} from '~/lib/types/leetcode';

export async function fetchLeetcodeQuestionListMetadata(
  leetcodeQuestionFilters: LeetcodeQuestionFilters,
): Promise<LeetcodeQuestionMetadata[]> {
  try {
    const response = await axios.get(
      `${process.env.EXPO_PUBLIC_API_BASE_URL}/api/leetcode/metadata`,
      {
        headers: {
          'Content-Type': 'application/json',
        },
        params: {
          difficulty: leetcodeQuestionFilters.difficulty,
          topic_tag: leetcodeQuestionFilters.topicTag,
          page: leetcodeQuestionFilters.page,
        },
      },
    );

    return response.data.map((q: LeetcodeQuestionMetadata) => leetcodeQuestionMetadata.parse(q));
  } catch (error) {
    console.error('Error fetching question list metadata:', error);
    throw error;
  }
}

export async function fetchLeetcodeQuestionData(titleSlug: string): Promise<LeetcodeQuestion> {
  try {
    const response = await axios.get(`${process.env.EXPO_PUBLIC_API_BASE_URL}/api/leetcode`, {
      headers: {
        'Content-Type': 'application/json',
      },
      params: {
        title_slug: titleSlug,
      },
    });

    return leetcodeQuestion.parse(response.data);
  } catch (error) {
    console.error('Error fetching question data:', error);
    throw error;
  }
}

File: server/app/controllers/leetcode.py
import logging

import httpx
from fastapi import APIRouter, HTTPException

from app.models.leetcode import (
    Difficulty,
    LeetcodeQuestion,
    LeetcodeQuestionMetadata,
    TopicTag,
)
from app.services.questions import LeetcodeService

log = logging.getLogger(__name__)


class LeetcodeController:
    def __init__(self, service: LeetcodeService):
        self.router = APIRouter()
        self.service = service
        self.setup_routes()

    def setup_routes(self):
        router = self.router

        @router.get(
            "/metadata",
            response_model=list[LeetcodeQuestionMetadata],
        )
        async def fetch_leetcode_question_list_metadata(
            difficulty: Difficulty, topic_tag: TopicTag, page: int
        ) -> list[LeetcodeQuestionMetadata]:
            """Fetches the list of question metadata for the user to choose from, based on the provided difficulty and topic tag."""
            try:
                log.info("Fetching question list metadata...")
                response: list[LeetcodeQuestionMetadata] = (
                    await self.service.fetch_leetcode_question_list_metadata(
                        difficulty=difficulty, topic_tag=topic_tag, page=page
                    )
                )
                log.info(f"{len(response)} questions' metadata retrieved successfully.")
                return response
            except Exception as e:
                log.error(
                    "Unexpected error in questions controller while fetching list of question metadata: %s",
                    str(e),
                )
                raise HTTPException(status_code=httpx.codes.INTERNAL_SERVER_ERROR, detail=str(e))

        @router.get(
            "",
            response_model=LeetcodeQuestion,
        )
        async def fetch_leetcode_question(title_slug: str) -> LeetcodeQuestion:
            try:
                log.info("Fetching question data...")
                response: LeetcodeQuestion = await self.service.fetch_leetcode_question(
                    title_slug=title_slug
                )
                log.info(f"{response.title} data retrieved successfully.")
                return response
            except Exception as e:
                log.error(
                    "Unexpected error in questions controller while fetching question data: %s",
                    str(e),
                )
                raise HTTPException(status_code=httpx.codes.INTERNAL_SERVER_ERROR, detail=str(e))
